ext/dom/xml_document.c

```diff
@@ -0,0 +1,258 @@
+/*
+   +----------------------------------------------------------------------+
+   | Copyright (c) The PHP Group                                          |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 3.01 of the PHP license,      |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | https://www.php.net/license/3_01.txt                                 |
+   | If you did not receive a copy of the PHP license and are unable to   |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@php.net so we can mail you a copy immediately.               |
+   +----------------------------------------------------------------------+
+   | Authors: Niels Dossche <nielsdos@php.net>                            |
+   +----------------------------------------------------------------------+
+*/
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "php.h"
+#if defined(HAVE_LIBXML) && defined(HAVE_DOM)
+#include "php_dom.h"
+#include "namespace_compat.h"
+
+static bool check_options_validity(uint32_t arg_num, zend_long options)
+{
+	const zend_long VALID_OPTIONS = XML_PARSE_NOENT
+								  | XML_PARSE_DTDLOAD
+								  | XML_PARSE_DTDATTR
+								  | XML_PARSE_DTDVALID
+								  | XML_PARSE_NOERROR
+								  | XML_PARSE_NOWARNING
+								  | XML_PARSE_NOBLANKS
+								  | XML_PARSE_XINCLUDE
+								  | XML_PARSE_NSCLEAN
+								  | XML_PARSE_NOCDATA
+								  | XML_PARSE_NONET
+								  | XML_PARSE_PEDANTIC
+								  | XML_PARSE_COMPACT
+								  | XML_PARSE_HUGE
+								  | XML_PARSE_BIG_LINES;
+	if ((options & ~VALID_OPTIONS) != 0) {
+		zend_argument_value_error(2, "contains invalid flags (allowed flags: "
+									 "LIBXML_NOENT, "
+									 "LIBXML_DTDLOAD, "
+									 "LIBXML_DTDATTR, "
+									 "LIBXML_DTDVALID, "
+									 "LIBXML_NOERROR, "
+									 "LIBXML_NOWARNING, "
+									 "LIBXML_NOBLANKS, "
+									 "LIBXML_XINCLUDE, "
+									 "LIBXML_NSCLEAN, "
+									 "LIBXML_NOCDATA, "
+									 "LIBXML_NONET, "
+									 "LIBXML_PEDANTIC, "
+									 "LIBXML_COMPACT, "
+									 "LIBXML_PARSEHUGE, "
+									 "LIBXML_BIGLINES)");
+		return false;
+	}
+	return true;
+}
+
+/* Living spec never creates explicit namespace declaration nodes.
+ * They are only written upon serialization but never appear in the tree.
+ * So in principle we could just ignore them outright.
+ * However, step 10 in https://html.spec.whatwg.org/multipage/parsing.html#create-an-element-for-the-token
+ * requires us to have the declaration as an attribute available */
+static void dom_mark_namespaces_as_attributes_too(xmlDocPtr doc)
```

Comment:
Just to clarify, but why does it seem the DOM_XMLDocument class need to refer to something that looks from the HTML spec?

Reply:
For whatever BC reason, XML parsers encountering an xmlns attribute will interpret it both as a namespace declaration and an attribute. This does not matter in practice for XML because XML serialization will always output declarations.
In HTML, there are no namespace declarations in its serialization. However, if you import an XML node into an HTML document, and the XML node had a namespace at parse time, it will be introduced as both an attribute and a namespace declaration. So in the serialization you will see the xmlns output. However, if you make a namespace declaration after parsing, it does not have a double life as an attribute, and will therefore _not_ be included in its serialization.
This function marks namespaces as attributes too by setting a flag in the namespace declaration node. The function is called right after parsing.

Reply:
ACK